### PR TITLE
Reviews backend

### LINKED ISSUE #11: Bonus BE: Reviews
Create a reviews model
- can contain multiple reviews
- each review is 1-5 and an optional text
Implement
- `GET` all reviews for user
- `POST` make a "create review" route for testing reasons
Add the reviews to a user in whatever way you see fit

### PR DESCRIPTION
**Pull request by kevin-06-huang**
### What this PR does (required):
- implemented #11 
- built the backend for reviews tied to a specific user
- built model, controller, route (get reviews for a user and post
- not debugged with front end yet

### Any information needed to test this feature (required):
- install as per usual

### CODE REVIEW

File: server/controllers/review.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    review.save();
+
+    reviews = Profile.find({ userId: profileId }, { "reviews": 1 }).reviews;
+    if (reviews.length <= 10) {
+        reviews.push(review);
+    } else {
+        reviews.pop();
+        reviews.unshift(review);
+    };
+    Profile.updateOne({ userId: profileId }, {
```
Comment (ezeilo-su) [inline server/controllers/review.js:38]: Please `await` this `Promise` and eliminate callback

File: server/models/Profile.js [hunk 2]
```diff
@@ -42,7 +43,15 @@ const profileSchema = new mongoose.Schema({
     type: String,
     enum: ["pet_sitter", "pet_owner"],
     default: "pet_owner",
-  }
+  },
+  // most recent 10 reviews
```
Comment (ezeilo-su) [inline server/models/Profile.js:47]: Get rid of comment

**Review by sundayezeilo — CHANGES_REQUESTED** [review]
Hey @kevin-06-huang, great work on this one. Kindly have a look at the comments and address them.
Happy coding :smiley:

File: server/controllers/review.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+exports.getReviews = asyncHandler(async (req, res) => {
+    const { id } = req.user;
+    Review.find({ profileId: id })
+        .then((reviews) => {
+            res.status(200).json({
+                success: {
+                    reviews: reviews,
+                },
+            });
+        });
```
[the remark this answers is not in the document]
Comment (kevin-06-huang) [reply server/controllers/review.js:14]: Made change as per comment.

File: server/controllers/review.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+const Profile = require("../models/Profile");
+const asyncHandler = require("express-async-handler");
+
+exports.getReviews = asyncHandler(async (req, res) => {
+    const { id } = req.user;
+    Review.find({ profileId: id })
+        .then((reviews) => {
+            res.status(200).json({
+                success: {
+                    reviews: reviews,
```
[the remark this answers is not in the document]
Comment (kevin-06-huang) [reply server/controllers/review.js:11]: Ty! Made change as suggested.

File: server/controllers/review.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        star,
+        description
+    } = req.body;
+    review = new Review({
+        reviewerId,
+        profileId,
+        star,
+        description
+    });
+    review.save();
```
[the remark this answers is not in the document]
Comment (kevin-06-huang) [reply server/controllers/review.js:29]: Made change as suggested.

File: server/controllers/review.js (hunk 1 above)
Comment (kevin-06-huang) [reply server/controllers/review.js:38]: Made change as recommended.

**Comment by kevin-06-huang** [comment]
All comments addressed in commit https://github.com/hatchways/team-wintergreen/pull/82/commits/3412f5552d0108aebaeeab39184d9df8353d37b5.

File: server/models/Review.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
+const mongoose = require("mongoose");
+
+const ReviewSchema = new mongoose.Schema({
```
Comment (rajivtitus) [inline server/models/Review.js:3]: 👍🏽

File: server/models/Review.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  reviewerId: {
+    type: mongoose.Schema.Types.ObjectId,
+    required: true,
+    ref: "User",
+  },
+  profileId: {
```
Comment (rajivtitus) [inline server/models/Review.js:9]:
```suggestion
  revieweeId: {
```
Calling it 'profileId' seems misleading as we're storing a ref to the User model here

File: server/controllers/review.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
+const Review = require("../models/Review");
+const Profile = require("../models/Profile");
+const asyncHandler = require("express-async-handler");
+
+exports.getReviews = asyncHandler(async (req, res) => {
+    const { id } = req.user;
+    let reviews = await Review.find({ profileId: id });
```
Comment (rajivtitus) [inline server/controllers/review.js:7]:
```suggestion
    const reviews = await Review.find({ profileId: id });
```

File: server/controllers/review.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
+const Review = require("../models/Review");
+const Profile = require("../models/Profile");
+const asyncHandler = require("express-async-handler");
+
+exports.getReviews = asyncHandler(async (req, res) => {
+    const { id } = req.user;
```
Comment (rajivtitus) [inline server/controllers/review.js:6]: We will be fetching reviews for a sitter's profile and not the logged in user (considering passing the sitter's user/profile id as a route param)

File: server/controllers/review.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    res.status(200).json({ success: { reviews }});
+});
+exports.createReview = asyncHandler(async (req, res) => {
+    const {
+        reviewerId,
+        profileId,
+        star,
+        description
+    } = req.body;
+    let review = new Review({
```
Comment (rajivtitus) [inline server/controllers/review.js:17]:
```suggestion
    const review = new Review({
```

File: server/controllers/review.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+        reviewerId,
```
Comment (rajivtitus) [inline server/controllers/review.js:18]: We don't need to send the `reviewerId` using the request body, why don't we access it using the request object instead - `req.user.id`

File: server/controllers/review.js [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+        profileId,
+        star,
+        description
+    });
+    await review.save();
+
+    let reviews = await Profile.find({ userId: profileId }, { "reviews": 1 }).reviews;
```
Comment (rajivtitus) [inline server/controllers/review.js:25]:
```suggestion
    let sitterProfile = await Profile.findById(profileId, { "reviews": 1 });
```
`findById` of `findOne` seems more appropriate here

**Review by rajivtitus — COMMENTED** [review]
Good work on this Kevin! Just some comments for you to have a look and address 👍🏽

File: server/models/Review.js (hunk 7 above)
Comment (kevin-06-huang) [reply server/models/Review.js:9]: changed to suggested.

File: server/models/Review.js (hunk 6 above)
Comment (kevin-06-huang) [reply server/models/Review.js:3]: ty.

File: server/controllers/review.js (hunk 8 above)
Comment (kevin-06-huang) [reply server/controllers/review.js:7]: changed to suggested.

File: server/controllers/review.js (hunk 9 above)
Comment (kevin-06-huang) [reply server/controllers/review.js:6]: changed to using profileId param

File: server/controllers/review.js (hunk 10 above)
Comment (kevin-06-huang) [reply server/controllers/review.js:17]: changed to suggested.

File: server/controllers/review.js (hunk 11 above)
Comment (kevin-06-huang) [reply server/controllers/review.js:18]: changed to using req.user.id

File: server/controllers/review.js [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    if (reviews.length <= 10) {
+        reviews.push(review);
+    } else {
+        reviews.pop();
+        reviews.unshift(review);
+    };
+    let profile = await Profile.updateOne({ userId: profileId }, { $set: { reviews } })
```
[the remark this answers is not in the document]
Comment (kevin-06-huang) [reply server/controllers/review.js:32]: made changes to use save().

File: server/controllers/review.js (hunk 12 above)
Comment (kevin-06-huang) [reply server/controllers/review.js:25]: changed to findById

**Comment by kevin-06-huang** [comment]
All latest comments addressed in commit https://github.com/hatchways/team-wintergreen/pull/82/commits/83bc2d3fa858781d7696c95f8bb14ae35bbbbe57.

**Pull request closed without merging**